tokio-executor/src/park.rs

```diff
@@ -237,37 +238,56 @@ impl Inner {
     fn park(&self, timeout: Option<Duration>) -> Result<(), ParkError> {
         // If currently notified, then we skip sleeping. This is checked outside
         // of the lock to avoid acquiring a mutex if not necessary.
-        match self.state.compare_and_swap(NOTIFY, IDLE, Ordering::SeqCst) {
-            NOTIFY => return Ok(()),
-            IDLE => {},
-            _ => unreachable!(),
+        if self.state.compare_exchange(NOTIFY, IDLE, SeqCst, SeqCst).is_ok() {
+            return Ok(());
+        }
+
+        // If the duration is zero, then there is no need to actually block
+        if let Some(ref dur) = timeout {
+            if *dur == Duration::from_millis(0) {
+                return Ok(());
+            }
         }
 
         // The state is currently idle, so obtain the lock and then try to
         // transition to a sleeping state.
         let mut m = self.mutex.lock().unwrap();
 
         // Transition to sleeping
-        match self.state.compare_and_swap(IDLE, SLEEP, Ordering::SeqCst) {
-            NOTIFY => {
+        match self.state.compare_exchange(IDLE, SLEEP, SeqCst, SeqCst) {
+            Ok(_) => {}
+            Err(NOTIFY) => {
                 // Notified before we could sleep, consume the notification and
                 // exit
-                self.state.store(IDLE, Ordering::SeqCst);
+                self.state.swap(IDLE, SeqCst);
```

Comment:
Using swap as a store seems like overkill since the value is discarded. On x86 this will generate an unnecessary locked instruction. (ie, the std version should use store.)

Reply:
I believe that the version in std is incorrect. The `swap` is required to establish ordering with [this](https://github.com/tokio-rs/tokio/blob/06309e0cb67743ec223f45a2475f3ff95437beb2/tokio-executor/src/park.rs#L241) line in unpark. Without doing a "read" operation here, visibility cannot be acquired.

Reply:
I'm not sure I follow. I didn't check std, but I assume the `swap` here came from there.

Also, my understanding is that `SeqCst` should establish total ordering with other `SeqCst` loads and stores, so this can be a plain `store` and still be ordered with respect to `compare_exchange`. (The `std::atomic::Ordering` docs are irritatingly imprecise here though, so I don't know if I'm actually right - it seems to imply that `SeqCst` isn't meaningful for `store`.)

Reply:
@jsgf 

A store operation with `Release` ordering (or stronger) writes to memory and allows other future `load` operations to synchronize with it.

However, here we also want to perform a load operation at the same time (hence `swap`) and synchronize with whoever last wrote to this location (i.e. *acquire their writes to memory*).

Reply:
Doesn't `SeqCst` already guarantee that?

Reply:
No, because a `store` only has sequentially consistent `Release` semantics, but it isn't an `Acquire` operation. :)

Reply:
@jsgf the way to think about it, when it comes to **two** threads interacting, `SeqCst` is identical to `Acquire` / `Release`. `SeqCst` only adds additional guarantees when more than 2 threads interact.

Of course, that is per the spec. On x86 (I believe), SeqCst implies `mfence`, which is pretty strong.

Reply:
Relevant discussion: https://github.com/rust-lang/rfcs/pull/2503

Reply:
Hm, OK. Trouble is that on x86 loads and stores to the same location *are* totally ordered, and an "atomic" store can just be a regular store - whereas swap will be an atomic operation which is overkill here.

Reply:
It is possible that a fence might work, but that will make tsan grumpy :(

Reply:
Wow, there are a lot of atomic ops in these few lines (the mutex acquisition, the CAS, the store/swap).

I don’t follow the need for the `swap` either - isn’t the CAS right above already acquiring the writes done by a different thread?

Reply:
@jsgf, from an x86 codegen perspective, I don’t think a `SeqCst` store or swap is going to matter - it’s going to be some instruction that entails LOCK like behavior for both.  In fact, I see `xchg reg, mem` on the playground for both.

Reply:
> Isn't the read at line 257 sufficient?

I was about to ask the same. The read of `NOTIFY` with `SeqCst` is sufficient to establish synchronization with the thread that notified.

Of course release/acquire would be sufficient as well but I am not going to try playing that game [again](https://github.com/rust-lang/rust/pull/52349) ;)

Reply:
No, because a third thread can come along and `unpark` while the `park` fn is between 257 and 262. The bug is specifically that the `park` thread must acquire the memory from a third thread that calls unpark before the park thread hits line 262.

Reply:
(Continuing in https://github.com/rust-lang/rust/issues/53366)